src/main/java/org/openrewrite/java/logging/slf4j/StripToStringFromArguments.java

```diff
@@ -0,0 +1,59 @@
+package org.openrewrite.java.logging.slf4j;
+
+import org.openrewrite.ExecutionContext;
+import org.openrewrite.Preconditions;
+import org.openrewrite.Recipe;
+import org.openrewrite.TreeVisitor;
+import org.openrewrite.java.JavaIsoVisitor;
+import org.openrewrite.java.search.UsesMethod;
+import org.openrewrite.java.tree.Expression;
+import org.openrewrite.java.tree.J;
+import org.openrewrite.java.tree.TypeUtils;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class StripToStringFromArguments extends Recipe {
+    @Override
+    public String getDisplayName() {
+        return "Strip `toString()` from arguments";
+    }
+
+    @Override
+    public String getDescription() {
+        return "Remove `.toString()` from logger call arguments; SLF4J will automatically call `toString()` on an argument when not a string, and do so only if the log level is enabled.";
+    }
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getVisitor() {
+        return Preconditions.check(Preconditions.or(
+                        new UsesMethod<>("org.slf4j.Logger trace(..)"),
+                        new UsesMethod<>("org.slf4j.Logger debug(..)"),
+                        new UsesMethod<>("org.slf4j.Logger info(..)"),
+                        new UsesMethod<>("org.slf4j.Logger warn(..)"),
+                        new UsesMethod<>("org.slf4j.Logger error(..)")
+                ),
+                new JavaIsoVisitor<ExecutionContext>() {
+                    @Override
+                    public J.MethodInvocation visitMethodInvocation(J.MethodInvocation mi, ExecutionContext ctx) {
+                        final int firstFormatArgIndex = mi.getArguments().get(0).getType().toString() == "org.slf4j.Marker" ? 2 : 1;
+                        final List<Expression> newArguments = new ArrayList<>(mi.getArguments().subList(0, firstFormatArgIndex));
+                        for (int i = firstFormatArgIndex; i < mi.getArguments().size(); i++) {
+                            final Expression arg = mi.getArguments().get(i);
+                            Expression toAdd = arg;
+                            if (arg instanceof J.MethodInvocation) {
+                                J.MethodInvocation toStringInvocation = (J.MethodInvocation) arg;
+                                if (toStringInvocation.getSimpleName().equals("toString") &&
+                                        toStringInvocation.getSelect() != null &&
+                                        toStringInvocation.getSelect().getType() != null &&
```

Comment:
Could this ever be null? Can we remove this check?

Reply:
Type can in rare cases be null when there's issues with type attribution; it's unlikely though because those would then also cascade to the method invocation here, so it ought to be fairly safe to remove this here and rely on the TypeUtils method (or an alternative) below.